src/idealplastic.jl

```diff
@@ -1,73 +1,94 @@
-    time :: Float64 = zero(Float64)
-    strain :: SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
+module IdealPlasticModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, IS, ID, lame
+import ..integrate_material!  # for method extension
+
+export IdealPlastic, IdealPlasticDriverState, IdealPlasticParameterState, IdealPlaticVariableState
+
+@with_kw mutable struct IdealPlasticDriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
 end
 
-@with_kw struct IdealPlasticParameterState <: AbstractMaterialState
-    youngs_modulus :: Float64 = zero(Float64)
-    poissons_ratio :: Float64 = zero(Float64)
-    yield_stress :: Float64 = zero(Float64)
+@with_kw struct IdealPlasticParameterState{T <: Real} <: AbstractMaterialState
+    youngs_modulus::T = zero(T)
+    poissons_ratio::T = zero(T)
+    yield_stress::T = zero(T)
 end
 
-@with_kw struct IdealPlasticVariableState <: AbstractMaterialState
-    stress :: SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
-    plastic_strain :: SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
-    cumeq :: Float64 = zero(Float64)
-    jacobian :: SymmetricTensor{4,3} = zero(SymmetricTensor{4,3,Float64})
+@with_kw struct IdealPlasticVariableState{T <: Real} <: AbstractMaterialState
+    stress::Symm2{T} = zero(Symm2{T})
+    plastic_strain::Symm2{T} = zero(Symm2{T})
+    cumeq::T = zero(T)
+    jacobian::Symm4{T} = zero(Symm4{T})
 end
 
-@with_kw mutable struct IdealPlastic <: AbstractMaterial
-    drivers :: IdealPlasticDriverState = IdealPlasticDriverState()
-    ddrivers :: IdealPlasticDriverState = IdealPlasticDriverState()
-    variables :: IdealPlasticVariableState = IdealPlasticVariableState()
-    variables_new :: IdealPlasticVariableState = IdealPlasticVariableState()
-    parameters :: IdealPlasticParameterState = IdealPlasticParameterState()
-    dparameters :: IdealPlasticParameterState = IdealPlasticParameterState()
+@with_kw mutable struct IdealPlastic{T <: Real} <: AbstractMaterial
+    drivers::IdealPlasticDriverState{T} = IdealPlasticDriverState{T}()
+    ddrivers::IdealPlasticDriverState{T} = IdealPlasticDriverState{T}()
+    variables::IdealPlasticVariableState{T} = IdealPlasticVariableState{T}()
+    variables_new::IdealPlasticVariableState{T} = IdealPlasticVariableState{T}()
+    parameters::IdealPlasticParameterState{T} = IdealPlasticParameterState{T}()
+    dparameters::IdealPlasticParameterState{T} = IdealPlasticParameterState{T}()
 end
 
-function integrate_material!(material::IdealPlastic)
+"""
+    integrate_material!(material::IdealPlastic)
+
+Ideal plastic material: no hardening. The elastic region remains centered on the
+origin, and retains its original size.
+"""
+function integrate_material!(material::IdealPlastic{T}) where T <: Real
```

Comment:
In the English literature the "_ideal plastic_" term is actually "**perfect plastic**". The "ideal plastic" term is a direct translation from the Finnish teaching materials :-). Perhaps we should rename it to `PerfectPlastic` to ring more bells.

Reply:
Thanks, didn't know that. Let's change it.